src/blocks/sound/pipewire.rs

```diff
@@ -0,0 +1,643 @@
+                link_output_node,
+                link_input_node,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct Data {
+    pub nodes: HashMap<u32, Node>,
+    pub links: HashMap<u32, Link>,
+}
+
+#[derive(Default)]
+struct NodeProxies {
+    proxies_t: HashMap<u32, NodeProxy>,
+    listeners: HashMap<u32, Vec<Box<dyn Listener>>>,
+}
+
+impl NodeProxies {
+    fn add_node_proxy(
+        &mut self,
+        node_proxy: NodeProxy,
+        node_listener: NodeListener,
+        proxies: &std::rc::Rc<RefCell<NodeProxies>>,
+    ) -> u32 {
+        let listener_spe = Box::new(node_listener);
+
+        let proxy = node_proxy.upcast_ref();
+        let proxy_id = proxy.id();
+
+        let proxies_weak = Rc::downgrade(proxies);
+
+        let listener = proxy
+            .add_listener_local()
+            .removed(move || {
+                if let Some(proxies) = proxies_weak.upgrade() {
+                    proxies.borrow_mut().remove(proxy_id);
+                }
+            })
+            .register();
+
+        self.add_proxy_t(proxy_id, node_proxy, listener_spe);
+        self.add_proxy_listener(proxy_id, listener);
+
+        proxy_id
+    }
+
+    fn add_proxy_t(&mut self, proxy_id: u32, node_proxy: NodeProxy, listener: Box<dyn Listener>) {
+        self.proxies_t.insert(proxy_id, node_proxy);
+        self.listeners.entry(proxy_id).or_default().push(listener);
+    }
+
+    fn add_proxy_listener(&mut self, proxy_id: u32, listener: ProxyListener) {
+        self.listeners
+            .entry(proxy_id)
+            .or_default()
+            .push(Box::new(listener));
+    }
+
+    fn remove(&mut self, proxy_id: u32) {
+        self.proxies_t.remove(&proxy_id);
+        self.listeners.remove(&proxy_id);
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) enum EventKind {
+    NodeUpdate,
+    NodeParamUpdate,
+    NodeStateUpdate,
+    LinkUpdate,
+    NodeRemoval,
+    LinkRemoval,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) enum CommandKind {
+    Mute(u32, bool),
+    SetVolume(u32, u32),
+}
+
+impl CommandKind {
+    fn execute(self, client: &Client, proxies: Rc<RefCell<NodeProxies>>) {
+        use CommandKind::*;
+        match self {
+            SetVolume(id, volume) => {
+                if let Some(node) = client.data.lock().unwrap().nodes.get(&id) {
+                    if let (Some(channels), Some(proxy)) = (
+                        node.channels,
+                        proxies.borrow_mut().proxies_t.get(&node.proxy_id),
+                    ) {
+                        let volume = volume as f32 / NORMAL;
+                        let volume = volume * volume * volume;
+
+                        let mut pod_data = Vec::new();
+                        let mut pod_builder = PodBuilder::new(&mut pod_data);
+
+                        // let mut frame: ::std::mem::MaybeUninit<libspa::sys::spa_pod_frame> = ::std::mem::MaybeUninit::uninit();
```

Comment:
This comment is the equivalent of the `builder_add!` invocation below, but that requires a new feature (that's enabled by the patch in `Cargo.toml`)